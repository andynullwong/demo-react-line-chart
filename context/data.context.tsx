"use client";

import BtcAddressBalance from "@/types/btcAddressBalance.type";
import { ReactNode, createContext, useContext, useState } from "react";

type ParsedBtcBalance = Papa.ParseResult<BtcAddressBalance>;

export type DataContextType = {
  data: ParsedBtcBalance;
  setData: React.Dispatch<React.SetStateAction<ParsedBtcBalance>>;
};

const DataContext = createContext<DataContextType>({} as DataContextType);

export const DataContextProvider = ({ children }: { children: ReactNode }) => {
  const [data, setData] = useState({} as ParsedBtcBalance);

  return (
    <DataContext.Provider value={{ data, setData }}>
      {children}
    </DataContext.Provider>
  );
};

export const useDataContext = () => useContext(DataContext);
