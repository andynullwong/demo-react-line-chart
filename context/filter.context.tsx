"use client";

import ChartRange from "@/types/chartRange.type";
import { ReactNode, createContext, useContext, useState } from "react";

type FilterContextType = {
  filter: ChartRange;
  setFilter: React.Dispatch<React.SetStateAction<ChartRange>>;
};

const FilterContext = createContext<FilterContextType>({} as FilterContextType);

export const FilterContextProvider = ({
  children,
}: {
  children: ReactNode;
}) => {
  const [filter, setFilter] = useState(ChartRange.ALL);

  return (
    <FilterContext.Provider value={{ filter, setFilter }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
