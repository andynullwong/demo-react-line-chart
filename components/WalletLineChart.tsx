import { useDataContext } from "@/context/data.context";
import { useFilterContext } from "@/context/filter.context";
import ChartRange from "@/types/chartRange.type";
import {
  getDayMonthFromDateString,
  getYearFromDateString,
  numberToMillions,
} from "@/utils/formatNumbers";
import {
  findIndexOfMax,
  getDateRangeMax,
  walletChartCache,
} from "@/utils/walletUtil";
import Spinner from "@/components/Spinner";
import { parseISO } from "date-fns";
import React from "react";
import {
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";

const WalletLineChart = () => {
  const { data } = useDataContext();
  const { filter } = useFilterContext();

  if (!data?.data?.length) return <Spinner />;

  if (!walletChartCache.has(filter)) {
    const idx = findIndexOfMax(
      data.data.map((d) => parseISO(d.Time)),
      getDateRangeMax(filter)
    );
    walletChartCache.set(filter, idx);
  }

  const filteredData =
    filter === ChartRange.ALL
      ? data.data
      : data.data.slice(walletChartCache.get(filter)! + 1);

  const tickFormatter =
    filter === ChartRange.ALL
      ? getYearFromDateString
      : getDayMonthFromDateString;

  return (
    <ResponsiveContainer width="100%" height={480}>
      <LineChart
        width={800}
        height={300}
        data={filteredData}
        margin={{
          top: 5,
          right: 30,
          left: 20,
          bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis
          dataKey="Time"
          interval={Math.floor(filteredData.length / (2 * 6))}
          tickFormatter={tickFormatter}
        />
        <YAxis tickFormatter={numberToMillions} />
        <Tooltip
          formatter={(value) => numberToMillions(value.valueOf() as number)}
        />
        <Legend verticalAlign="top" />
        <Line
          dataKey=">1K"
          legendType="plainline"
          dot={false}
          stroke="#3F00FF"
          strokeWidth={2}
        />
        <Line
          dataKey=">10K"
          legendType="plainline"
          dot={false}
          stroke="#FFC300"
          strokeWidth={2}
        />
        <Line
          dataKey=">100K"
          legendType="plainline"
          dot={false}
          stroke="#FF0000"
          strokeWidth={2}
        />
        <Line
          dataKey=">1M"
          legendType="plainline"
          dot={false}
          stroke="#dda0dd"
          strokeWidth={2}
        />
        <Line
          dataKey=">10M"
          legendType="plainline"
          dot={false}
          stroke="#ff00ff"
          strokeWidth={2}
        />
      </LineChart>
    </ResponsiveContainer>
  );
};

export default WalletLineChart;
