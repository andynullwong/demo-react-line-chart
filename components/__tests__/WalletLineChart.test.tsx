import React from "react";
import { render } from "@testing-library/react";
import WalletLineChart from "@/components/WalletLineChart";
import { useDataContext } from "@/context/data.context";
import mockJson from "./__mocks__/data.json";

jest.mock("../../context/data.context", () => ({
  useDataContext: jest.fn(),
}));

describe("WalletLineChart", () => {
  it("renders with mocked data", () => {
    (useDataContext as jest.Mock).mockReturnValue({
      data: mockJson
    } as any);

    const { asFragment } = render(<WalletLineChart />);
    expect(asFragment()).toMatchSnapshot();
  });
});
