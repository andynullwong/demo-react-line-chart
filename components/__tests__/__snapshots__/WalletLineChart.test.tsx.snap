// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletLineChart renders with mocked data 1`] = `
<DocumentFragment>
  <div
    class="absolute right-1/2 bottom-1/2  transform translate-x-1/2 translate-y-1/2 "
  >
    <div
      class="border-t-transparent border-solid animate-spin  rounded-full border-blue-400 border-8 h-64 w-64"
    />
  </div>
</DocumentFragment>
`;
